crates/oxc_ast/src/serialize.rs

```diff
@@ -225,6 +227,71 @@ impl ESTree for NullLiteralRaw<'_> {
     }
 }
 
+/// Serializer for `value` field of `StringLiteral`.
+///
+/// Handle when `lossy` flag is set, indicating the string contains lone surrogates.
+#[ast_meta]
+#[estree(
+    ts_type = "string",
+    raw_deser = "
+        const lossy = DESER[bool](POS_OFFSET.lossy);
+        (THIS.lossy && THIS.raw !== null) ? (0, eval)(THIS.raw) : DESER[Atom](POS_OFFSET.value)
+    "
+)]
+pub struct StringLiteralValue<'a, 'b>(pub &'b StringLiteral<'a>);
+
+impl ESTree for StringLiteralValue<'_, '_> {
+    fn serialize<S: Serializer>(&self, serializer: S) {
+        let lit = self.0;
+        #[expect(clippy::if_not_else)]
+        if !lit.lossy {
+            lit.value.serialize(serializer);
+        } else {
+            self.serialize_lossy(serializer);
+        }
+    }
+}
+
+impl StringLiteralValue<'_, '_> {
+    #[cold]
+    #[inline(never)]
+    fn serialize_lossy<S: Serializer>(&self, serializer: S) {
+        // String contains lone surrogates
+        let lit = self.0;
+        let raw =
+            lit.raw.expect("`StringLiteral` with `lossy` flag set must have `raw` field populated");
+        let raw = raw.as_str();
+        let quote = raw.as_bytes().first().copied().unwrap();
+        let raw_unquoted = &raw[1..raw.len() - 1];
+        let raw_with_quotes_escaped = if quote == b'"' {
+            // String was in double quotes in original source, so it's valid JSON
+            Cow::Borrowed(raw_unquoted)
+        } else {
+            // String was in single quotes in original source.
+            // We need to replace any `"` characters with `\"`, and then it's valid JSON.
+            // We *don't* escape `"` characters preceded by an odd number of `\`s,
+            // because they're already escaped.
+            // We could make this more performant, but it should be vanishingly rare to hit this path anyway.
+            let mut encoded = Vec::with_capacity(raw_unquoted.len());
+            let mut has_slash = false;
+            for &b in raw_unquoted.as_bytes() {
+                if b == b'\\' {
+                    has_slash = !has_slash;
+                } else if b == b'"' && !has_slash {
+                    encoded.push(b'\\');
+                }
+                encoded.push(b);
+            }
```

Comment:
Is `has_slash` correct? It looks like it's counting all previous occurrence instead of preceding sequence. 
For example, `'\uD83D"'` is edge case and it doesn't escape to `"\uD83D\""`?

(I know this is extremely rare edge cases, but I wonder if we can add extra manual test case since test262 didn't have this one).

Reply:
You're right! I neglected to reset `has_slash = false` after a non-slash character. Fixed now.

Anyway, this is superceded by #10041.

As far as tests, it would be a pain to add an ESTree test, because it'd complicate the test runner. But we can add parser tests in #10041. There are some other new tests we need for that PR anyway.